DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,308 @@
+
+When handling an error, don't log it only to pass it up the call stack. Logging and passing the received error up the call stack typically results in lots verbose and unhelpful logging about the same error.
+
+If appropriate, add context by wrapping the error using `fmt.Errorf` or equivilent. Only add context that is known by the scope of your function and do not _assume_ an error represents something unless you can make a positive assertion. This avoids embedding knowledge that makes re-use and refactors harder. For example:
+
+```go
+// foo knows bar can return errZero so can perform an explicit check.
+func foo() error {
+    err := bar(0)
+    if errors.Is(err, errZero) {
+        return fmt.Errorf("ambiguous: %v", err)
+    }
+
+    if err != nil {
+        return err
+    }
+}
+
+var errZero = errors.New("zero value")
+
+func bar(i int) error {
+    if i > 5 {
+        return errors.New("out of bounds")
+    }
+
+    if i == 0 {
+        return errZero
+    }
+}
+```
+
+### Boolean parameters
+
+Boolean parameters are widely considered an antipattern. Context is hugely important when assessing a boolean parameter but in general they represent an on/off switch for behavior. The behavioral on/off switch creates hard to understand and maintain logic that often has poor re-use and is an indicator that a piece of code violates the SRP. 
+
+Instead, consider structuring your code to have dynamically constructable behavior logic. This allows you to isolate booleans to construction logic, a distinct concern, and construct the behavior you require for your context. 
+
+An example of code designed to be dynamically constructed may include the [decorator](https://refactoring.guru/design-patterns/decorator) pattern. It may use a [builder](https://refactoring.guru/design-patterns/builder) or an [abstract factory](https://refactoring.guru/design-patterns/abstract-factory) with a builder to construct the logic using boolean flag configuration read from CLI flags. There are many more combinations and, generally, you want to consider your problem space and design a piece of code that makes sense for your problem first.
+
+### Concurrency
+
+Concurrency introduces harder to rationalize execution relative to serialized programs and is easy to get wrong. Concurrency doesn't necessarily speed up a program either. With these points in mind, prefer synchronous APIs instead leaving concurrency to the caller. Not only does this eliminate a whole host of problems, it ensures your APIs are predictable and easier to use.
+
+Dave Cheney has made some excellent [comments](https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#concurrency) on concurrency in Go.
+
+#### Channels
+
+If you find the need for channels prefer accepting them as parameters. Accepting channels as a parameter allows consumers to specify channel properties, such as buffered or unbuffered, creating a more flexible API. 
```

Comment:
Clarification: channels for input, channels for input or both?
I would agree with channels for input, I would disagree with channels for output. At least as general advice.

Reply:
Both. Even more so for output. As the doc notes, allowing consumers to pass the channel lets them control buffering which, as a consumer, is hugely important if you have an API experiencing bursts while you try to provide a more consistent throughput from your API. 

We should also add a note about conditionality of the channel.

Reply:
If you feel strongly about this, go ahead, I'm ok with it 🙂 

However, I don't necessarily agree this should always be the answer. What you are proposing does in fact offer a more flexible API but in some cases the price to pay would be a more complicated implementation.

To you example, burst smoothing can be easily implemented (kind of as a decorator) even if the concurrent code method returns a non buffered channel. I'm not saying it has to be that way, just that imo this is more a tradeoff topic than a clear rule of thumb.